File: wallet/src/main/java/chatox/wallet/repository/BalanceRepository.java
package chatox.wallet.repository;

import chatox.wallet.model.Balance;
import chatox.wallet.model.Currency;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;
import java.util.Optional;

public interface BalanceRepository extends JpaRepository<Balance, String> {
    List<Balance> findByUserId(String userId);
    Optional<Balance> findByUserIdAndCurrency(String userId, Currency currency);
    List<Balance> findByUserIdAndCurrencyIn(String userId, List<Currency> currencies);
}

File: user/src/main/kotlin/chatox/user/repository/BalanceRepository.kt
package chatox.user.repository

import chatox.user.domain.Balance
import chatox.user.domain.Currency
import org.springframework.data.mongodb.repository.ReactiveMongoRepository
import reactor.core.publisher.Mono

interface BalanceRepository : ReactiveMongoRepository<Balance, String> {
    fun findByUserIdAndCurrency(userId: String, currency: Currency): Mono<Balance>
}